### PR TITLE
Fix overload operator-> for CallExpression

### PR DESCRIPTION
**Pull request by peckto**

### CODE REVIEW

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt [hunk 1]
```diff
@@ -398,6 +398,22 @@ open class SymbolResolver(ctx: TranslationContext) : ComponentPass(ctx) {
         val callee = call.callee
         val language = call.language
 
+        // Check for a possible overloaded operator->
+        if (
+            callee.language is HasOperatorOverloading &&
+                callee is MemberExpression &&
+                callee.operatorCode == "->" &&
+                callee.base.type !is PointerType
```
Comment (oxisto) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt:406]: We pretty much have the same code for member access. Is it possible to extract it unto a common function?

**Comment by peckto** [comment]
Can someone explain to me, why the codecov/patch QA fails? I don't get it.

**Comment by oxisto** [comment]
> Can someone explain to me, why the codecov/patch QA fails? I don't get it.

It is probably failing because you are not reaching all branch targets in the line (yellow).

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt (hunk 1 above)
Comment (peckto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt]: yes, good point. Fixed

**Comment by peckto** [comment]
> > Can someone explain to me, why the codecov/patch QA fails? I don't get it.
> 
> It is probably failing because you are not reaching all branch targets in the line (yellow).

The red one seems to be gone now. Dont know about the yellow one, but it does not block the CI ^^

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt [hunk 2]
```diff
@@ -335,13 +342,8 @@ open class SymbolResolver(ctx: TranslationContext) : ComponentPass(ctx) {
...
         // Check for a possible overloaded operator->
-        if (
-            reference.language is HasOperatorOverloading &&
-                reference is MemberExpression &&
-                reference.operatorCode == "->" &&
-                reference.base.type !is PointerType
-        ) {
-            val result = resolveOperator(reference)
+        if (isOverloadedOperator(reference, "->")) {
+            val result = resolveOperator(reference as MemberExpression)
```
Comment (peckto) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt:346]: Is it possible to get rid of the `as MemberExpression` which is ensured by the `isOverloadedOperator` function? The linter seems not to get this.

Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt:346]: Yes with a fancy Kotlin magic called contracts :)
```
contract { returns() implies (ex is MemberExpression) }
```

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt [hunk 3]
```diff
@@ -398,6 +400,17 @@ open class SymbolResolver(ctx: TranslationContext) : ComponentPass(ctx) {
         val callee = call.callee
         val language = call.language
 
+        // Check for a possible overloaded operator->
+        if (isOverloadedOperator(callee, "->")) {
```
Comment (oxisto) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt:404]: I actually meant extracting the whole resolveOpreator, best viable, etc. stuff. isn't this 1:1 the same in both cases?

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt [hunk 4]
```diff
@@ -322,6 +322,13 @@ open class SymbolResolver(ctx: TranslationContext) : ComponentPass(ctx) {
         }
     }
 
+    private fun isOverloadedOperator(ex: Expression, operatorCode: String): Boolean {
```
Comment (oxisto) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt:325]: a) some small kdoc would be nice
b) I would prefer that the name of the function reflects that this only checks for overloaded member expressions.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt (hunk 3 above)
Comment (peckto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt]: Yes, basically, I was just thinking about the type update in the `resolveMember` case.
I now solved it as optional return type.
I don't see where the type should be used in the `handleCallExpression` case. Do you?

Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt]: Looks good, just add some documentation now and we're done :)

Comment (peckto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt]: Added doc for function `resolveOverloadedArrowOperator`

**Review by oxisto — APPROVED** [review]

**Pull request merged**